docs/async/async-in-depth.md

```diff
@@ -0,0 +1,121 @@
+# Async In Depth
+
+By [Phillip Carter](https://github.com/cartermp)
+
+Writing I/O-Bound or CPU-Bound asynchronous code is simple using the `async` and `await` keywords, but what's happening under the covers is fairly complex.  The two key objects involved are `Task` and `Task<T>`.
+
+## Task and Task<T>
+
+Tasks are constructs used to represent operations running in the background.
+
+*   `Task` represents a single operation which does not return a value.
+*   `Task<T>` represents a single operation which returns a value of type `T`.
+
+Tasks are awaitable, meaning that using `await` will allow your application or service to perform useful work while the task is running by yielding control to its caller until the task is done. If you’re using `Task<T>`, the `await` keyword will additionally “unwrap” the value returned when the Task is complete.
+
+It’s important to reason about Tasks as abstractions of work happening in the background, and _not_ an abstraction over multithreading. In fact, unless explicitly started on a new thread via `Task.Run`, a Task will start on the current thread and delegate work to the Operating System.
+
+## Deeper Dive into Tasks for an I/O-Bound Operation
+
+Here’s a 10,000 foot view of what happens with a typical async I/O call:
+
+```csharp
+public async Task<string> GetHtml()
+{
+	var client = new HttpClient();
+	
+	// Execution is still synchronous here!
+	// The task handle "getHtmlTask" represents the active HTTP request.
+	var getHtmlTask = client.GetStringAsync("http://www.dotnetfoundation.org");
```

Comment:
I'm thinking about if that's worth mentioning.  It's generally a bad thing to mix mutability and concurrency because you then cause non-deterministic side effects.  I call that out in the C# tutorial on async already, but it might be worth mentioning that you can technically do that, but you really should have referentially transparent asynchronous calls.